scanner/src/main/kotlin/scanners/fossid/FossIdNamingProvider.kt

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (C) 2021 Bosch.IO GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.scanner.scanners.fossid
+
+import java.time.LocalDateTime
+import java.time.format.DateTimeFormatter
+
+import org.ossreviewtoolkit.utils.log
+
+/**
+ * This class provides names for projects and scans when the FossID scanner creates them, following a given pattern.
+ * If one or both patterns is null, a default naming convention is used.
+ *
+ * [namingScanPattern] and [namingProjectPattern] are patterns describing the name using variables, e.g. "$var1_$var2".
+ * Variable values are given in the map [namingConventionVariables].
+ *
+ * There also are builtin variables. Builtin variables are are prefixed in the pattern with "#" e.g. "$var1_#builtin".
+ * Available builtin variables:
+ * * **projectBaseCode**: The name of the project (i.e. the part of the URL before .git).
+ */
+class FossIdNamingProvider(
+    private val namingProjectPattern: String?,
+    private val namingScanPattern: String?,
+    private val namingConventionVariables: Map<String, String>
+) {
+    fun createProjectName(baseProjectCode: String): String = namingProjectPattern?.let {
+        replaceNamingConventionVariables(namingProjectPattern, baseProjectCode, namingConventionVariables)
+    } ?: baseProjectCode
+
+    fun createScanName(baseProjectCode: String): String = if (namingScanPattern == null) {
+        val formatter = DateTimeFormatter.ofPattern("'$baseProjectCode'_yyyyMMdd_HHmmss")
+        formatter.format(LocalDateTime.now())
+    } else {
+        replaceNamingConventionVariables(namingScanPattern, baseProjectCode, namingConventionVariables)
+    }
+
+    /**
+     * Replace the naming convention variables with their values. Used for projects and scans.
+     */
+    private fun replaceNamingConventionVariables(
+        namingConventionPattern: String, baseProjectCode: String,
+        namingConventionVariables: Map<String, String>
+    ): String {
+        log.info { "Parametrizing the name with naming=$namingConventionPattern, baseProjectCode=$baseProjectCode" }
+        var replacedName = namingConventionPattern
```

Comment:
I guess, this can be simplified. You can create a map that contains both the built-in and the user-defined variables. Then the replacement can be done in a single step. You can also use `fold`, e.g.:
```
val replacedName = namingConventionVariables.fold(namingConventionPattern) { str, variable ->
  str.replace("\$${variable.key}", variable.value)
}
```

Reply:
1. the built-ins are prefix with "#" so you cannot replace variables and build-ins in one step...
2. I don't find `.fold` for `Map`. Do you mean `foldTo` for `MutableMap` ?

Reply:
at 1: This depends how you organize the keys in the map and the replacement logic, but I won't insist.
at 2: You can use `Map.entries().fold()`.

Reply:
It's ok...I reworked the code to do it in one operation.

Thanks for the suggestions !